x/staking/abci.go

```diff
@@ -50,7 +50,12 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) []abci.ValidatorUpdate {
 			k.DeleteByKey(ctx, iterator.Key())
 		}
 
-		return k.BlockValidatorUpdates(ctx)
+		// defer TODO should update epochNumber after epoch finish
+		// This won't affect slashing module since slashing Endblocker run before staking module
+
+		return k.EpochValidatorUpdates(ctx)
 	}
-	return []abci.ValidatorUpdate{}
+
+	// run block validator updates for slashed, jailed validators
+	return k.BlockValidatorUpdates(ctx)
```

Comment:
EpochValidatorUpdates will encompass BlockValidatorUpdates, right?  Because what if a validator gets jailed in the same block as EpochUpdates?

Reply:
It will be detected by EpochValidatorUpdates, EpochValidatorUpdates just do every validator update at the end of that block.